presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,749 @@
+                pagesRemoved += handleInfo.getPageCount();
+                bytesRemoved += handleInfo.getBytes();
+
+                startPage.set(0);
+            }
+            else {
+                pagesAcknowledged += sequenceId - range.lowerEndpoint() - startPage.get();
+                startPage.set(toIntExact(sequenceId - range.lowerEndpoint()));
+                break;
+            }
+        }
+
+        totalBufferedPages.addAndGet(-pagesRemoved);
+        totalBufferedBytes.addAndGet(-bytesRemoved);
+        totalPagesRemaining.addAndGet(-pagesAcknowledged);
+
+        return pagesAcknowledged;
+    }
+
+    private synchronized void acknowledgePagesFromMemory(long sequenceId, long startingSequenceId)
+    {
+        checkState(startingSequenceId >= currentMemorySequenceId.get(), "Invalid startingSequenceId for memory pages");
+        int pagesToRemove = toIntExact(sequenceId - startingSequenceId);
+        checkArgument(pagesToRemove <= pages.size(), "Invalid sequenceId");
+
+        long bytesRemoved = 0;
+        for (int i = 0; i < pagesToRemove; i++) {
+            SerializedPage removedPage = pages.remove();
+            bytesRemoved += removedPage.getRetainedSizeInBytes();
+            currentMemorySequenceId.incrementAndGet();
+        }
+
+        totalBufferedPages.addAndGet(-pagesToRemove);
+        totalBufferedBytes.addAndGet(-bytesRemoved);
+        totalPagesRemaining.addAndGet(-pagesToRemove);
+    }
+
+    @Override
+    public void abort(OutputBufferId bufferId)
+    {
+        checkArgument(bufferId.getId() == outputBufferId.getId(), "Invalid outputBufferId");
+        destroy();
+    }
+
+    @Override
+    public void setNoMorePages()
+    {
+        PendingRead pendingRead;
+        synchronized (this) {
+            state.compareAndSet(OPEN, NO_MORE_PAGES);
+            state.compareAndSet(NO_MORE_BUFFERS, FLUSHING);
+            noMorePages.set(true);
+
+            pendingRead = this.pendingRead;
+            this.pendingRead = null;
+        }
+
+        if (pendingRead != null) {
+            processPendingRead(pendingRead);
+        }
+    }
+
+    @Override
+    public void destroy()
+    {
+        PendingRead pendingRead;
+        synchronized (this) {
+            if (state.setIf(FINISHED, oldState -> !oldState.isTerminal())) {
+                for (HandleInfo info : handleInfoQueue) {
+                    info.removeFile();
+                }
+                pages.clear();
+                handleInfoQueue.clear();
+                noMorePages.set(true);
+                totalBufferedPages.set(0);
+                totalBufferedBytes.set(0);
+                totalPagesRemaining.set(0);
+            }
+            pendingRead = this.pendingRead;
+        }
+
+        if (pendingRead != null) {
+            pendingRead.completeResultFutureWithEmpty();
+        }
+    }
+
+    @Override
+    public void fail()
+    {
+        // TODO: clean up memory with a close method
```

Comment:
Is that TODO will be implemented in this PR since this is a WIP?

Reply:
Yes, this PR will take care of this.